---
title: "JK_chip_rnaseq"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r libraries}
library("GenomicRanges")
library("GenomicFeatures")
library("rtracklayer")
library("AnnotationHub")
library("ChIPpeakAnno")
library("magrittr")
library("dplyr")
library("biomaRt")

ts<-format(Sys.time(), "%a_%b_%d_%Y_%H%M")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```

# Download annotations and MGI data
```{r}
#download.file("ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M16/gencode.vM16.annotation.gff3.gz",destfile="gencode.vM16.annotation.gff3.gz")
#R.utils::gunzip("gencode.vM16.annotation.gff3.gz")

#download.file("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz",destfile="mm10.blacklist.bed.gz")
#R.utils::gunzip("mm10.blacklist.bed.gz")

bl_mm10<-import("mm10.blacklist.bed")
seqlevelsStyle(bl_mm10)<-"NCBI"

txdb<-makeTxDbFromGFF("gencode.vM16.annotation.gff3",format="gff3")
seqlevelsStyle(txdb)<-"NCBI"
gencode<-exonsBy(txdb,by="gene")
transcripts<-transcripts(txdb)
keytypes(txdb)
conversion<-AnnotationDbi::select(txdb,columns="GENEID",keytype="TXNAME",keys=transcripts$tx_name)
transcripts$ensembl<-substr(conversion[match(transcripts$tx_name,conversion$TXNAME),"GENEID"],1,18)

#txdb still doesn't incorporate mgi names so we will still pull those from biomaRt
listMarts(host="www.ensembl.org")
ensembl_91<-useMart(biomart="ENSEMBL_MART_ENSEMBL",host="www.ensembl.org", 
                    dataset="mmusculus_gene_ensembl")

mgi <- getBM(attributes = c("ensembl_gene_id","mgi_symbol"), filters = "ensembl_gene_id", values = unique(transcripts$ensembl), mart = ensembl_91)
transcripts$mgi<-mgi[match(transcripts$ensembl,mgi$ensembl_gene_id),]$mgi_symbol
names(transcripts)<-transcripts$tx_name

transcripts[grep("^Bcor$",transcripts$mgi)]
```

# Define Functions
```{r define_functions,eval=T}

wls<- function(x) { paste0("chr",seqnames(x),":",start(x),"-",end(x))}

broadPeakToGRanges<-function(file,name="peak", ...) {
  x <- read.table(file,stringsAsFactors=F, ...)
  gr <-GRanges(seqnames=x$V1, ranges = IRanges(start=x$V2, end=x$V3),
               strand="*", score=x$V5, e=x$V7,p=x$V8,q=x$V9)
  names(gr)<-paste0(name,"_",formatC(1:length(gr),width=5,format="d",flag="0"))
  return(gr)
}


narrowPeakToGRanges<-function(file,name="peak", ...) {
  x <- read.table(file,stringsAsFactors=F, ...) 
  gr <-GRanges(seqnames=x$V1, ranges = IRanges(start=x$V2, end=x$V3),
               strand="*", score=x$V5, e=x$V7,summit=x$V10)
  names(gr)<-paste0(name,"_",formatC(1:length(gr),width=5,format="d",flag="0"))
  return(gr)
}

annotate_peaks<-function(gr) {
  seqlevelsStyle(gr)<-"NCBI"
  gr<-gr[!gr %over% bl_mm10]
  
  gr_anno<-annotatePeakInBatch(gr, AnnotationData=transcripts, output="both", maxgap=5000L)
  seqlevelsStyle(gr_anno)<-"NCBI"
  
  
  gr_anno$symbol<-transcripts$mgi[match(gr_anno$feature,transcripts$tx_name)]
  
  
  x<-mcols(gr_anno) %>% as.data.frame %>%
  dplyr::filter(symbol!="") %>% 
  dplyr::group_by(peak) %>%
  summarize(gene=paste(unique(symbol),collapse=";")) %>% as.data.frame

  
  gr$genes<-x[match(names(gr),x$peak),"gene"]
  gr$wls<-wls(gr)
  gr<-gr[with(gr,order(-score))]
  
  #return chacter vector of peaks associated with each ENSEMBL ID
  return(gr)
  
}

annotate_genes<-function(df,gr) {
  seqlevelsStyle(gr)<-"NCBI"
  gr<-gr[!gr %over% bl_mm10]
  
  gr_anno<-annotatePeakInBatch(gr, AnnotationData=transcripts, output="both", maxgap=5000L)
  seqlevelsStyle(gr_anno)<-"NCBI"
  

  
  gr_anno$ensembl<-transcripts[match(gr_anno$feature,transcripts$tx_name),]$ensembl
  gr_anno$wls<-wls(gr_anno)
    
  x<-mcols(gr_anno) %>% as.data.frame %>%
  #dplyr::mutate(peak=paste0(peak,"(",wls,")")) %>% 
  dplyr::group_by(ensembl) %>%
  summarize(peak=paste(unique(wls),collapse=";"),maxscore=max(score)) 
  
   # Rownames supercede column named "ensembl"
  if (substr(rownames(df)[1],1,7)=="ENSMUSG") {idx <<- match(rownames(df),x$ensembl)
  } else if ("ensembl" %in% colnames(df)) {idx<<-match(df$ensembl,x$ensembl)
  } else {idx<<-rep(NA,nrow(df))}
  
  df$maxscore<-x[idx,]$maxscore
  df$peak<-x[idx,]$peak
  
  return(df)
  
}
```

# Download Raw Data
```{r}
Annotate
```{r annotate,eval=T}

#import
length(th17n_Bcor_peaks<-narrowPeakToGRanges(url("https://s3.msi.umn.edu/bardwell/th17_Bcor_WT_113017_peaks.narrowPeak"),name="th17")) 
length(th17b_Bcor_peaks<-broadPeakToGRanges(url("https://s3.msi.umn.edu/jenkinsm/th17_bcor_083117_peaks.broadPeak"),name="th17")) 
length(e95_Bcor_peaks<-narrowPeakToGRanges(url("https://s3.msi.umn.edu/bardwell/e95_Bcor_WT_113017_peaks.narrowPeak"),name="e95"))
length(e85_Bcor_peaks<-narrowPeakToGRanges(url("https://s3.msi.umn.edu/bardwell/e85_Bcor_WT_113017_peaks.narrowPeak"),name="e85"))
length(h3_Bcor_peaks<-narrowPeakToGRanges(url("https://s3.msi.umn.edu/bardwell/crisprH3_Bcor_WT_113017_peaks.narrowPeak"),name="h3"))
length(cgi<-narrowPeakToGRanges(url("https://s3.msi.umn.edu/kybam/cpg_mm10.bed"),name="cgi",skip=1)) 
seqlevelsStyle(cgi)<-"NCBI"
#length(cgi<-narrowPeakToGRanges("cpg_mm10.bed",name="cgi",skip=1)) 

#save Peaklists for Offline Analysis  
save(th17n_Bcor_peaks,th17b_Bcor_peaks,e95_Bcor_peaks,e85_Bcor_peaks,h3_Bcor_peaks,cgi,file="peaklists.rdata")
#load("peaklists.rdata")

head(e95_Bcor_peaks<-annotate_peaks(e95_Bcor_peaks))
head(e85_Bcor_peaks<-annotate_peaks(e85_Bcor_peaks))
head(h3_Bcor_peaks<-annotate_peaks(h3_Bcor_peaks))
head(th17b_Bcor_peaks<-annotate_peaks(th17b_Bcor_peaks))

length(th17_specific_noPla<-   th17b_Bcor_peaks [!(th17b_Bcor_peaks %over% e95_Bcor_peaks )])
length(th17_specific_noPla_noCGI <-   th17_specific_noPla [!(th17_specific_noPla %over% cgi )])
length(th17_specific<-   th17b_Bcor_peaks [!(th17b_Bcor_peaks %over% e95_Bcor_peaks | th17b_Bcor_peaks %over% cgi) ])
View(as.data.frame(th17_specific))

length(th17_specific<-th17_Bcor_peaks[!(th17_Bcor_peaks %over% e95_Bcor_peaks |th17_Bcor_peaks %over% e85_Bcor_peaks | th17_Bcor_peaks %over% h3_Bcor_peaks) ])
th17_specific<-annotate_peaks(th17_specific)
View(as.data.frame(th17_specific))

```

#Test out the annotate_genes function with Teng's Data
```{r}
combine.subset<-read.csv("/Users/gearh006/Documents/Notebook/GCD Consultant/bardwell/teng/combine_subset_w_terms_Tue_Dec_12_2017_1359.csv",stringsAsFactors = F)

combine.subset<-annotate_genes(df=combine.subset,e95_Bcor_peaks)
View(combine.subset)
write.csv(combine.subset,file=paste0("combine_subset_w_terms_",ts,".csv"),quote=F,row.names = F)
```

