---
title: "JY177 RNA-Seq"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Setup
## Load Libraries
```{r libraries,eval=T,results='hide',message=FALSE,warning=FALSE}
library("magrittr")
library("dplyr")
library("stringr")
library("ggplot2")
library("DESeq2")
library("ComplexHeatmap")
library("GenomicFeatures")
library("rtracklayer")
library("GenomicRanges")
library("ChIPpeakAnno")
library("ComplexHeatmap")
library("gridExtra")
library("goseq")
library("regioneR")

source("custom_plotPCA.R")
ts<-format(Sys.time(), "%a_%b_%d_%Y_%H%M")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```

## Download Annotations & Counts
```{r gencode_annotations,eval=F}
#Download one of the following
#download.file("ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M16/gencode.vM16.annotation.gff3.gz",destfile="gencode.vM16.annotation.gff3.gz")
#R.utils::gunzip("gencode.vM16.annotation.gff3.gz")

#download.file("ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M16/gencode.vM16.basic.annotation.gff3.gz",destfile="gencode.vM16.basic.annotation.gff3.gz")
#R.utils::gunzip("gencode.vM16.basic.annotation.gff3.gz")

# Kundaje Blacklist
#download.file("http://mitra.stanford.edu/kundaje/akundaje/release/blacklists/mm10-mouse/mm10.blacklist.bed.gz",destfile="mm10.blacklist.bed.gz")
#R.utils::gunzip("mm10.blacklist.bed.gz")

#point to file downloaded above
gff3_file<-"../../R_RESOURCES/gencode.vM16.annotation.gff3"

txdb<-makeTxDbFromGFF(gff3_file,format="gff3")
  seqlevelsStyle(txdb)<-"NCBI"

#get symbols and biotypes from gff3 file that is not in txdb
M16<-import(gff3_file,format="gff3")
seqlevelsStyle(M16)<-"NCBI"
symbols<-as.data.frame(mcols(M16))
symbols<-symbols[symbols$type=="gene",c("gene_id","gene_name","gene_type")]
symbols$chr<-as.character(seqnames(M16[match(symbols$gene_id,M16$gene_id)]))
symbols$gene_id<-substr(symbols$gene_id,1,18)
table(symbols$chr)

#add symbol & biotype to a transcripts GRanges object

transcripts<-transcripts(txdb,columns=c("gene_id","tx_name"))
transcripts$gene_id<-substr(unlist(transcripts$gene_id),1,18)
names(transcripts)<-transcripts$tx_name
mcols(transcripts)<-data.frame(gene_id=transcripts$gene_id,stringsAsFactors = F)
transcripts$symbol<-symbols$gene_name[match(transcripts$gene_id,symbols$gene_id)]
transcripts$biotype<-symbols$gene_type[match(transcripts$gene_id,symbols$gene_id)]

#promoters
prom<-promoters(txdb,upstream=2000,downstream=200,use.names=T)
seqlevelsStyle(prom)<-"NCBI"
all.equal(rownames(prom),rownames(transcripts))
prom$gene_id<-transcripts$gene_id
prom$symbol<-transcripts$symbol

#exons
exons<-exonsBy(txdb,"tx",use.names=TRUE)
#names(exons)<-substr(names(exons),1,18)
#any more from MGI table?
#MGI<-read.table("gencode.vM16.metadata.MGI",header=F,stringsAsFactors = F)
#summary(idx<-match(names(transcripts),MGI$V1))
#transcripts$MGI<-MGI[idx,"V2"]
#table(transcripts$symbol==transcripts$MGI,useNA="always")
#transcripts[which(!transcripts$symbol==transcripts$MGI)]
#transcripts$MGI[transcripts$symbol==transcripts$MGI)]

bl_mm10<-import("mm10.blacklist.bed")
seqlevelsStyle(bl_mm10)<-"NCBI"

save(symbols,transcripts,exons,prom,bl_mm10,file="symbols_M16.rdata")

#load(url("https://s3.msi.umn.edu/jenkinsm/featureCounts_Gencode-M16_q0_Sun_Jan_21_2018_1932.rdata"))
#load(url("https://s3.msi.umn.edu/gearhart/mgi91.rdata"))

# with PCR duplicates
#load(url("https://jenkinsm.s3.msi.umn.edu/featureCounts_Gencode-M16_q0_Tue_Feb_06_2018_0947.rdata"))

#with duplicates removed 
#load(url("https://jenkinsm.s3.msi.umn.edu/featureCounts_Gencode-M16_q0_Tue_Mar_20_2018_0751.rdata"))
download.file("https://jenkinsm.s3.msi.umn.edu/featureCounts_Gencode-M16_q0_Tue_Mar_20_2018_0751.rdata",destfile = "featureCounts_Gencode-M16_q0_Tue_Mar_20_2018_0751.rdata")

#load(url("https://s3.msi.umn.edu/gearhart/symbols_M16.rdata"))

#MGI GO annotations
#download.file("http://www.informatics.jax.org/downloads/reports/gene_association.mgi.gz",destfile="../../R_RESOURCES/gene_association.mgi.gz")
#R.utils::gunzip("../../R_RESOURCES/gene_association.mgi.gz")
#download.file("http://www.informatics.jax.org/downloads/reports/go_terms.mgi",destfile="../../R_RESOURCES/go_terms.mgi")

```

## Load GO Annotations from  MGI
```{r load_go,eval=F}

mgi_go<-readr::read_delim("../../R_RESOURCES/gene_association.mgi",delim="\t",skip=24,col_names=F)

colnames(mgi_go)<-c("DB","DB Object ID","Mgi Symbol","Qualifier","GO ID","DB:eEference","Evidence Code",
                    "WithorFrom","Aspect","DB Object Name","DB Object Syn","DB Object Type","Taxon","Date","Assigned By",
                    "Annotation Extension","Gene Product Form ID")
table(mgi_go$`Evidence Code`)
#high quality?  
#dim(mgi_go<-mgi_go[mgi_go$`Evidence Code` %in% c("EXP","IDA","IEP","IGI","IMP","IPI","ISS","TAS"),])

summary(idx<-match(mgi_go$`Mgi Symbol`,mcols(transcripts)$symbol))
mgi_go$ensembl<-mcols(transcripts)$gene_id[idx]

dim(mgi_go<-mgi_go[!is.na(mgi_go$ensembl),c("ensembl","GO ID")])

#subset to BP
mgi_goterms<-readr::read_delim("../../R_RESOURCES//go_terms.mgi",delim="\t",col_names=F)
colnames(mgi_goterms)<-c("Category","GO ID","Description")
table(mgi_goterms$Category)
dim(mgi_BP_goterms<-mgi_goterms[mgi_goterms$Category=="Biological Process","GO ID"])
dim(mgi_go<-mgi_go[mgi_go$`GO ID` %in% mgi_BP_goterms$`GO ID`,])

mgi_go[grep("ENSMUSG00000044770",mgi_go$ensembl),]
#dim(mgi_go<-mgi_go[mgi_go$ensembl %in% expressed_genes,])


#remove duplicated entries (possibly from multiple lines of evidence)
dim(mgi_go<-mgi_go[!duplicated(mgi_go),])

mgi_go.list<-split(mgi_go$`GO ID`,mgi_go$ensembl)

mgi_go.listByTerm<-split(mgi_go$ensembl,mgi_go$`GO ID`)
mgi_go.listByGene<-split(mgi_go$`GO ID`,mgi_go$ensembl)

#add term length to mgi_goterms
mgi_go.listByTerm<-split(mgi_go$ensembl,mgi_go$`GO ID`)
term_length<-sapply(mgi_go.listByTerm,length)
mgi_goterms$length<-term_length[mgi_goterms$`GO ID`]

as.data.frame(mgi_goterms[mgi_goterms$`GO ID` %in% mgi_go.list[[symbols[grep("^Tbx21$",symbols$gene_name),"gene_id"]]],])
as.data.frame(mgi_goterms[mgi_goterms$`GO ID` %in% mgi_go.list[[symbols[grep("^Scml4$",symbols$gene_name),"gene_id"]]],])




#View(mgi_goterms[mgi_goterms$`GO ID` %in% mgi_go.list[["ENSMUSG00000024837"]],])
#View(mgi_goterms[mgi_goterms$`GO ID` %in% mgi_go.list[["ENSMUSG00000050397"]],])

symbols[symbols$gene_id %in% mgi_go[grep("GO:0030217",mgi_go$`GO ID`),"ensembl"]$ensembl,] #T Cell Differentiotion

#bias.data
#bd<-sum(width(reduce(ens87)))
bd<-rev_stranded_counts$annotation$Length
names(bd)<-substr(rev_stranded_counts$annotation$GeneID,1,18)
bd["ENSMUSG00000001444"] #tbet


save(bd,mgi_go.list,mgi_goterms,mgi_go.listByGene,mgi_go,file="mgi_go.rdata")
```


## Load Data
```{r load_data,eval=T}
load("featureCounts_Gencode-M16_q0_Tue_Mar_20_2018_0751.rdata")
load("symbols_M16.rdata")
load("mgi_go.rdata")
```

## Define Functions
```{r functions,eval=T}
gg_plotCounts<-function(x="ENSMUSG00000028150",d=dds_subset) {
  if (substr(x,1,7)=="ENSMUSG") {
   # title<-mgi[grep(x,mgi$ensembl_gene_id),"mgi_symbol"]
    title<-symbols[grep(x,symbols$gene_id),"gene_name"]
  } else {
    title<-x
    #x<-mgi[grep(paste0("^",title,"$"),mgi$mgi_symbol),"ensembl_gene_id"]
    x<-symbols[grep(paste0("^",title,"$"),symbols$gene_name),"gene_id"]
  }

  plotCounts(d,x,intgroup=c("project","celltype","genotype"),returnData=T) %>%
    ggplot(aes(x=genotype, y=count,color=project)) +
    geom_point(position=position_jitter(w=0.1,h=0)) + ggtitle(paste0(x,":  ",title)) +
    stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
                 geom = "crossbar", width = 0.35,color="blue",size=0.2) +
    expand_limits(x=0, y = 0) + xlab("") +
    facet_grid(~celltype) +
    theme(legend.position="none") +
    theme_bw()
}



wls<- function(x) { paste0("chr",seqnames(x),":",start(x),"-",end(x))}

broadPeakToGRanges<-function(file,name="peak", ...) {
  x <- read.table(file,stringsAsFactors=F, ...)
  gr <-GRanges(seqnames=x$V1, ranges = IRanges(start=x$V2, end=x$V3),
               strand="*", score=x$V5, e=x$V7,p=x$V8,q=x$V9)
  names(gr)<-paste0(name,"_",formatC(1:length(gr),width=5,format="d",flag="0"))
  return(gr)
}


narrowPeakToGRanges<-function(file,name="peak", ...) {
  x <- read.table(file,stringsAsFactors=F, ...) 
  gr <-GRanges(seqnames=x$V1, ranges = IRanges(start=x$V2, end=x$V3),
               strand="*", score=x$V5, e=x$V7,summit=x$V10)
  names(gr)<-paste0(name,"_",formatC(1:length(gr),width=5,format="d",flag="0"))
  return(gr)
}

annotate_peaks<-function(gr,tx=transcripts,cutoff=10000) {
  seqlevelsStyle(gr)<-"NCBI"
  gr<-gr[!gr %over% bl_mm10]
  
  gr_anno<-annotatePeakInBatch(gr, AnnotationData=tx, output="both", maxgap=1000L)
  gr_anno<-subset(gr_anno,shortestDistance<=cutoff)
  seqlevelsStyle(gr_anno)<-"NCBI"
  
  
  gr_anno$symbol<-tx$symbol[match(gr_anno$feature,names(tx))]
  
  
  x<-mcols(gr_anno) %>% as.data.frame %>%
  dplyr::filter(symbol!="") %>% 
  dplyr::group_by(peak) %>%
  summarize(gene=paste(unique(symbol),collapse=";")) %>% as.data.frame

  
  gr$genes<-x[match(names(gr),x$peak),"gene"]
  gr$wls<-wls(gr)
  gr<-gr[with(gr,order(-score))]
  
  #return chacter vector of peaks associated with each ENSEMBL ID
  return(gr)
  
}

peaks_to_genes<-function(gr,tx=transcripts,cutoff=10000) {
  seqlevelsStyle(gr)<-"NCBI"
  gr<-gr[!gr %over% bl_mm10]
  
  gr_anno<-annotatePeakInBatch(gr, AnnotationData=tx, output="both", maxgap=1000L)
  gr_anno<-subset(gr_anno,shortestDistance<=cutoff)
  gr_anno$ensembl<-tx[match(gr_anno$feature,names(tx))]$gene_id

  #return chacter vector of ENSEMBL IDs
  return(unique(gr_anno$ensembl))
}

annotate_genes<-function(df,gr,tx=transcripts,cutoff=10000) {
  seqlevelsStyle(gr)<-"NCBI"
  gr<-gr[!gr %over% bl_mm10]
  
  gr_anno<-annotatePeakInBatch(gr, AnnotationData=tx, output="both", maxgap=1000L)
  gr_anno<-subset(gr_anno,shortestDistance<=cutoff)
  seqlevelsStyle(gr_anno)<-"NCBI"
  

  
  gr_anno$gene_id<-tx[match(gr_anno$feature,names(tx)),]$gene_id
  gr_anno$wls<-wls(gr_anno)
    
  x<-mcols(gr_anno) %>% as.data.frame %>%
  #dplyr::mutate(peak=paste0(peak,"(",wls,")")) %>% 
  dplyr::group_by(gene_id) %>%
  summarize(peak=paste(unique(wls),collapse=";"),maxscore=max(score)) 
  
   # Rownames supercede column named "ensembl"
  if (substr(rownames(df)[1],1,7)=="ENSMUSG") {idx <<- match(rownames(df),x$gene_id)
  } else if ("ensembl" %in% colnames(df)) {idx<<-match(df$ensembl,x$gene_id)
  } else {idx<<-rep(NA,nrow(df))}
  
  df$maxscore<-x[idx,]$maxscore
  df$peak<-x[idx,]$peak
  
  df<-df[with(df,order(-maxscore)),]
  
  return(df)
  
}

geneToGOIDs<-function(x) {
#  terms<-mgi_go.list[[symbols[grep(x,symbols$gene_name),"gene_id"]]]
#  as.data.frame(mgi_goterms[mgi_goterms$`GO ID` %in% terms,])
  paste(mgi_goterms$Description[mgi_goterms$`GO ID` %in% mgi_go.listByGene[[x]]],collapse = ";")
}
#geneToGOIDs("ENSMUSG00000001444")

listGO<-function(goid,dg=degs) {
  print(as.data.frame(mgi_goterms[grep(goid,mgi_goterms$`GO ID`),1:3]))
  tg<-as.data.frame(mgi_go[grep(goid,mgi_go$`GO ID`),"ensembl"])
  tg<-tg[tg$ensembl %in% expressed_genes,,drop=F]
  cat(paste0("There are ",nrow(tg)," expressed genes in this category.\n"))
  idx<-match(tg$ensembl,symbols$gene_id)
  tg$symbol<-symbols[idx,"gene_name"]
  tg$deg<-dg[tg$ensembl]
  tg<-tg[tg$deg==1,]
  cat (paste0("The ",nrow(tg)," differentially expressed genes are:\n"))
  return(tg)
}
```





## Build dds object
```{r build_dds,eval=T}
#cbind(as.data.frame(apply(rev_stranded_counts$counts,2,sum)),as.data.frame(apply(rev_stranded_unfiltered$counts,2,sum)))

coldata<-as.data.frame(apply(rev_stranded_counts$counts,2,sum))
colnames(coldata)<-"total_counts"
dim(coldata)

coldata$sample_name<-rownames(coldata) %>% 
  gsub("\\.\\.\\.Project_033p2\\.GRCm38\\.","",.) %>% 
  gsub("GRCm38\\.","",.) %>% 
    gsub("\\.Aligned\\.sortedByCoord\\.out\\.dedup\\.unique\\.bam","",.)  %>% 
  gsub("\\.Aligned\\.sortedByCoord\\.out\\.bam","",.)  %>% 
   gsub("\\.dedup\\.unique\\.bam","",.) %>% 
  gsub("\\.bam","",.) %>% 
  gsub("expB_KO_Other_idx15","tother_KO_repB_P033_Pool2_Index15",.) %>%
  gsub("expB_KO_Tfh_idx16","tfh_KO_repB_P033_Pool2_Index16",.) %>%
  gsub("expB_KO_Tfh.Th17_idx14","tfhth17_KO_repB_P033_Pool2_Index14",.) %>%
  gsub("expB_KO_Th17_idx13","th17_KO_repB_P033_Pool2_Index13",.) %>%
  gsub("expB_WT_Other_idx11","tother_WT_repB_P033_Pool2_Index11",.) %>%
  gsub("expB_WT_Tfh_idx12","tfh_WT_repB_P033_Pool2_Index12",.) %>%
  gsub("expB_WT_Tfh.Th17_idx10","tfhth17_WT_repB_P033_Pool2_Index10",.) %>%
  gsub("expB_WT_Th17_idx9","th17_WT_repB_P033_Pool2_Index9",.) %>%
  gsub("expD_KO_Other_idx7","tother_KO_repD_P033_Pool2_Index7",.) %>%
  gsub("expD_KO_Tfh_idx8","tfh_KO_repD_P033_Pool2_Index8",.) %>%
  gsub("expD_KO_Tfh.Th17_idx6","tfhth17_KO_repD_P033_Pool2_Index6",.) %>%           
  gsub("expD_KO_Th17_idx5","th17_KO_repD_P033_Pool2_Index5",.) %>%
  gsub("expD_WT_Other_idx3","tother_WT_repD_P033_Pool2_Index3",.) %>%
  gsub("expD_WT_Tfh_idx4","tfh_WT_repD_P033_Pool2_Index4",.) %>%
  gsub("expD_WT_Tfh.Th17_idx2","tfhth17_WT_repD_P033_Pool2_Index2",.) %>%           
  gsub("expD_WT_Th17_idx1","th17_WT_repD_P033_Pool2_Index1",.) 

dds<-DESeqDataSetFromMatrix(rev_stranded_counts$counts,colData = coldata,design = ~1)
rownames(colData(dds)) <- colData(dds)$sample_name
rownames(dds)<-substr(rownames(dds),1,18)
mcols(dds)$basepairs<-rev_stranded_counts$annotation$Length
#mcols(dds)$mgi_symbol<-mgi[match(rownames(dds),mgi$ensembl_gene_id),]$mgi_symbol
mcols(dds)$symbol<-symbols[match(rownames(dds),symbols$gene_id),]$gene_name

colData(dds)$celltype<-factor(as.character(str_split_fixed(colData(dds)$sample_name,"_", 6)[,1]),levels=c("th17","tfh","tfhth17","tother"))
colData(dds)$genotype<-factor(as.character(str_split_fixed(colData(dds)$sample_name,"_", 6)[,2]),levels=c("WT","KO"))
colData(dds)$rep<-as.character(str_split_fixed(colData(dds)$sample_name,"_", 6)[,3])
colData(dds)$project<-factor(as.character(str_split_fixed(colData(dds)$sample_name,"_", 6)[,4]),levels=c("P033","P041"))
colData(dds)$pool<-as.character(str_split_fixed(colData(dds)$sample_name,"_", 6)[,5])
colData(dds)$index<-as.character(str_split_fixed(colData(dds)$sample_name,"_", 6)[,6])

dds<-estimateSizeFactors(dds)
#eliminate samples with < 20% of reads
as.data.frame(colData(dds[,colData(dds)$sizeFactor < 0.20]))
as.data.frame(colData(dds_subset<-dds[,colData(dds)$sizeFactor > 0.20]))
dds_subset<-estimateSizeFactors(dds_subset)
design(dds_subset)<- ~project +celltype + genotype  #take the bias due to the project into account
```

## Make a table of Mean FPKM and mean count values for each condition
```{r fpkm_table,eval=T}
fpkm<-fpkm(dds_subset,robust=TRUE)

f_mean <- fpkm %>% as.data.frame %>% 
#  dplyr::filter(mgi_symbol %in% c(goi,goi2)) %>%
  tibble::rownames_to_column(var="ensembl") %>%
  tidyr::gather(sample,fpkm,-ensembl) %>%
  dplyr::mutate(time=colData(dds_subset)[sample,]$celltype) %>%
  dplyr::mutate(gene=colData(dds_subset)[sample,]$genotype) %>%
  dplyr::mutate(sample=paste(time,gene,sep="_")) %>%
  dplyr::group_by(sample,ensembl) %>%
    dplyr::summarize(mean_fpkm=round(mean(fpkm),3)) %>%
    dplyr::select(sample,ensembl,mean_fpkm) %>%
    ungroup() %>%
  tidyr::spread(sample,mean_fpkm) %>% as.data.frame()

rownames(f_mean)<-f_mean$ensembl
f_mean<-f_mean[,-1]
colnames(f_mean)<-paste0("fpkm_",colnames(f_mean))

count_mean <- counts(dds_subset,normalized=T) %>% as.data.frame %>% 
  tibble::rownames_to_column(var="ensembl") %>% 
 # dplyr::filter(ensembl %in% goiDF$gene_id) %>% 
  tidyr::gather(sample,counts,-ensembl) %>% 
  dplyr::mutate(time=colData(dds_subset)[sample,]$celltype) %>%
  dplyr::mutate(gene=colData(dds_subset)[sample,]$genotype) %>%
  dplyr::mutate(sample=paste(time,gene,sep="_")) %>% 
  dplyr::group_by(sample,ensembl) %>%
    dplyr::summarize(mean_count=round(mean(counts),2)) %>%
    dplyr::select(sample,ensembl,mean_count) %>%
    ungroup() %>%
  tidyr::spread(sample,mean_count) %>% as.data.frame()

  rownames(count_mean)<-count_mean$ensembl
count_mean<-count_mean[,-1]
colnames(count_mean)<-paste0("count_",colnames(count_mean))

#Export Data for GEO
norm_fpkm<-round(as.data.frame(fpkm(dds_subset,robust=TRUE)),3)
colnames(norm_fpkm)<-paste0(colData(dds_subset)$celltype,"_",colData(dds_subset)$genotype,"_",colData(dds_subset)$rep)
apply(norm_fpkm,2,sum)
norm_fpkm$`Gene Symbol` <-symbols[match(rownames(norm_fpkm),symbols$gene_id),"gene_name"]
dim(norm_fpkm<-tibble::rownames_to_column(norm_fpkm,var="Ensembl Gene ID"))
dim(norm_fpkm<-norm_fpkm[,c(1,38,2:37)])
write.csv(norm_fpkm,file=paste0("Normalized_FPKM_",ts,".csv"),quote=F,row.names = F)

raw_counts<-as.data.frame(counts(dds_subset,normalized=F))
colnames(raw_counts)<-paste0(colData(dds_subset)$celltype,"_",colData(dds_subset)$genotype,"_",colData(dds_subset)$rep)
apply(raw_counts,2,sum)
raw_counts$`Gene Symbol` <-symbols[match(rownames(raw_counts),symbols$gene_id),"gene_name"]
dim(raw_counts<-tibble::rownames_to_column(raw_counts,var="Ensembl Gene ID"))
dim(raw_counts<-raw_counts[,c(1,38,2:37)])
write.csv(raw_counts,file=paste0("Raw_Counts_",ts,".csv"),quote=F,row.names = F)

```

# EDA
## PCA plots
```{r pca,eval=F}
# log2 transform
g1<-plotPCA.DESeqTransform(normTransform(dds_subset),intgroup=c("celltype","genotype"),ntop=500,dims=c(1,2)) + theme_bw() + theme(legend.position = "none")
g2<-plotPCA.DESeqTransform(normTransform(dds_subset),intgroup=c("celltype","genotype"),ntop=500,dims=c(3,1)) + theme_bw() + theme(legend.position = "none")
g3<-plotPCA.DESeqTransform(normTransform(dds_subset),intgroup=c("celltype","genotype"),ntop=500,dims=c(3,2)) + theme_bw() + theme(legend.position = "none")
g4<-plotPCA.DESeqTransform(normTransform(dds_subset),intgroup=c("celltype","genotype"),ntop=500,dims=c(2,3)) + theme_bw() 
grid.arrange(g1,g3,g_legend(g4),g2, ncol=2,top="All Genes Log2 Transformd PCA")


# rlog transform
rlog_dds_subset<-rlog(dds_subset,blind=T)
g1r<-plotPCA.DESeqTransform(rlog_dds_subset,intgroup=c("celltype","genotype"),ntop=500,dims=c(1,2)) + theme_bw() + theme(legend.position = "none")
g2r<-plotPCA.DESeqTransform(rlog_dds_subset,intgroup=c("celltype","genotype"),ntop=500,dims=c(3,1)) + theme_bw() + theme(legend.position = "none")
g3r<-plotPCA.DESeqTransform(rlog_dds_subset,intgroup=c("celltype","genotype"),ntop=500,dims=c(3,2)) + theme_bw() + theme(legend.position = "none")
g4r<-plotPCA.DESeqTransform(rlog_dds_subset,intgroup=c("celltype","genotype"),ntop=500,dims=c(2,3)) + theme_bw() 
grid.arrange(g1r,g3r,g_legend(g4r),g2r, ncol=2,top="All Genes Rlog Transformd PCA")

pdf(file=paste0("pca_plot_",ts,".pdf"),width=6,height=6)
plotPCA.DESeqTransform(rlog_dds_subset,intgroup=c("celltype","genotype"),ntop=500,dims=c(1,2)) + theme_bw() + ggtitle("Rlog Transformed PCA")
dev.off()

g1r+scale_shape_manual(values=c(1,20))+theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"))

#+scale_color_manual(values=c("green","blue","red","black"))

ggplot_build(g1r)$data

# focus on T cell genes only?
tcell_genes<-as.data.frame(mgi_goterms[grep("T cell",mgi_goterms$Description),])
length(tcell_genes<-unique(as.character(unlist(mgi_go.listByTerm[tcell_genes$`GO ID`]))))

dim(dds_tcell<-dds_subset[rownames(dds_subset) %in% tcell_genes,])
plotPCA.DESeqTransform(normTransform(dds_tcell),intgroup=c("celltype","genotype"),ntop=Inf,dims=c(1,2)) + theme_bw() + ggtitle("T Cell Genes")
rlog_dds_tcell<-rlog(dds_tcell,blind=T)
plotPCA.DESeqTransform(rlog_dds_tcell,intgroup=c("celltype","genotype"),ntop=Inf,dims=c(1,2)) + theme_bw() + ggtitle("T Cell Genes - Rlog")

```

## Eigenvectors for PCA
```{r eigenvectors_for_pca,eval=F}
rlog_dds_subset.pca<-plotPCA.DESeqTransform(rlog_dds_subset,intgroup=c("celltype","genotype"),returnData=T)
nt.rot<-attr(rlog_dds_subset.pca,"rotation")
plot(rlog_dds_subset.pca[,c(1,2)],col=ifelse(stringr::str_detect(rownames(rlog_dds_subset.pca),"_KO_"),"red","black"),pch=16)
topN <- 5

#pc1
nt.rot1<-as.data.frame(nt.rot[order(abs(nt.rot[,1]),decreasing=TRUE),])[,1,drop=F]
head(left_join(tibble::rownames_to_column(nt.rot1,var="gene_id"),symbols,by="gene_id"),topN) 

#pc2
nt.rot2<-as.data.frame(nt.rot[order(abs(nt.rot[,2]),decreasing=TRUE),])[,2,drop=F]
head(left_join(tibble::rownames_to_column(nt.rot2,var="gene_id"),symbols,by="gene_id"),topN)

#pc3
nt.rot3<-as.data.frame(nt.rot[order(abs(nt.rot[,3]),decreasing=TRUE),])[,3,drop=F]
head(left_join(tibble::rownames_to_column(nt.rot3,var="gene_id"),symbols,by="gene_id"),topN)

```

## Make normalized FPKM table and heatmap of T cell genes
```{r fpkm_heatmap,eval=T}

goi<-list(th17=c("Rorc", "Ccr6", "Maf", "Il17a","Cxcr6"),
          tfh=c( "Bcl6", "Cxcr5", "Pdcd1", "Il4", "Il21"),
          tother=c( "Tbx21", "Ifng"))

goiDF<-symbols[symbols$gene_name %in% unlist(goi),]
rownames(goiDF)<-goiDF$gene_name
goiDF<-goiDF[unlist(goi),]
goiDF$celltype<-"x"
for (i in 1:nrow(goiDF)) {goiDF$celltype[i]<-names(goi)[grep(rownames(goiDF)[i],goi)]}
goiDF$celltype=factor(goiDF$celltype,levels=c("th17","tfh","tother"))

colnames(fpkm)<-paste(colData(dds_subset)$celltype,colData(dds_subset)$genotype,colData(dds_subset)$rep,sep="_")
#temp<-log2(fpkm[tcell_genes,]+0.1)
temp<-log2(fpkm[goiDF$gene_id,]+0.1)

rownames(temp)<-symbols[match(rownames(temp),symbols$gene_id),"gene_name"]
temp<-temp[,sort(colnames(temp))]


temp2<-as.matrix(t(scale(t(temp),center=T,scale=T)))

# Make Heatmap

col_fun = circlize::colorRamp2(c(min(temp2),0, max(temp2)), c("blue","white", "red"))

ht_th17 = Heatmap(temp2[,c(which(str_detect(colnames(temp2),"^th17_WT")),which(str_detect(colnames(temp2),"^th17_KO")))], 
                  column_title = "th17", rect_gp = gpar(col = "black", lwd = 0.5),
                  cluster_rows = TRUE, cluster_columns=FALSE,split = goiDF$celltype,
                  row_names_side = "left",col = col_fun, show_heatmap_legend = TRUE)

ht_tfth17 = Heatmap(temp2[,c(which(str_detect(colnames(temp2),"^tfhth17_WT")),which(str_detect(colnames(temp2),"^tfhth17_KO")))],
                    column_title = "tfhth17", rect_gp = gpar(col = "black", lwd = 0.5),
                    cluster_rows = TRUE, cluster_columns=FALSE,split = goiDF$celltype,
                    show_row_names = FALSE,col = col_fun,show_heatmap_legend = FALSE)
ht_tfh = Heatmap(temp2[,c(which(str_detect(colnames(temp2),"^tfh_WT")),which(str_detect(colnames(temp2),"^tfh_KO")))],
                 column_title = "tfh", rect_gp = gpar(col = "black", lwd = 0.5),
                 cluster_rows = TRUE, cluster_columns=FALSE,split = goiDF$celltype,
                 show_row_names=F,col = col_fun,show_heatmap_legend = FALSE)

ht_tother = Heatmap(temp2[,c(which(str_detect(colnames(temp2),"^tother_WT")),which(str_detect(colnames(temp2),"^tother_KO")))],
                    column_title = "tother", rect_gp = gpar(col = "black", lwd = 0.5),
                    cluster_rows = TRUE, cluster_columns=FALSE, split = goiDF$celltype,
                    show_row_names=F,row_dend_side = "right",col = col_fun,show_heatmap_legend = FALSE)

pdf(file=paste0("heatmap_",ts,".pdf"),width=8,height=8)
draw(ht_th17 + ht_tfth17 + ht_tfh + ht_tother,main_heatmap=1,split = goiDF$celltype,row_dend_side="right")
dev.off()
```



# Differential Expression
 
## Create a group variable that combines cell type and genotype to make individual contrasts for each cell type
```{r grouped_dds,eval=T}
dds_subset_grouped<-dds_subset
dds_subset_grouped$group <- factor(paste0(dds_subset_grouped$celltype, dds_subset_grouped$genotype))
design(dds_subset_grouped) <- ~project + group
dds_subset_grouped <- DESeq(dds_subset_grouped)
#resultsNames(dds_subset_grouped)

#th17
summary(res_global_th17_raw<-results(dds_subset_grouped,contrast=c("group","th17KO","th17WT"),alpha=0.05))
res_global_th17<-as.data.frame(res_global_th17_raw)
res_global_th17$symbol<-mcols(dds_subset_grouped)$symbol
stopifnot(all.equal(res_global_th17$symbol,symbols[match(rownames(res_global_th17),symbols$gene_id),"gene_name"]))
#View(as.data.frame(res_th17))
#nrow(res_global_th17<-subset(res_global_th17,abs(log2FoldChange) > log2(1.5) & padj <0.05 & baseMean > 10)) # subset to significant
#res_global_th17<-res_global_th17[with(res_global_th17,order(padj)),] #sort based on padj
#head(as.data.frame(res_global_th17),20)
#gg_plotCounts(rownames(res_global_th17[1,]))

#tfh
summary(res_global_tfh_raw<-results(dds_subset_grouped,contrast=c("group","tfhKO","tfhWT"),alpha=0.05))
res_global_tfh<-as.data.frame(res_global_tfh_raw)
res_global_tfh$symbol<-mcols(dds_subset_grouped)$symbol
stopifnot(all.equal(res_global_tfh$symbol,symbols[match(rownames(res_global_tfh),symbols$gene_id),"gene_name"]))
#nrow(res_global_tfh<-subset(res_global_tfh,abs(log2FoldChange) > log2(1.5) & padj <0.05 & baseMean > 10)) # subset to significant
#res_global_tfh<-res_global_tfh[with(res_global_tfh,order(padj)),] #sort based on padj
#head(as.data.frame(res_global_tfh),10)

#tfhth17
summary(res_global_tfhth17_raw<-results(dds_subset_grouped,contrast=c("group","tfhth17KO","tfhth17WT"),alpha=0.05))
res_global_tfhth17<-as.data.frame(res_global_tfhth17_raw)
res_global_tfhth17$symbol<-mcols(dds_subset_grouped)$symbol
stopifnot(all.equal(res_global_tfhth17$symbol,symbols[match(rownames(res_global_tfhth17),symbols$gene_id),"gene_name"]))
#nrow(res_global_tfhth17<-subset(res_global_tfhth17,abs(log2FoldChange) > log2(1.5) & padj <0.05 & baseMean > 10)) # subset to significant
#res_global_tfhth17<-res_global_tfhth17[with(res_global_tfhth17,order(padj)),] #sort based on padj
#head(as.data.frame(res_global_tfhth17),10)

#tother
summary(res_global_tother_raw<-results(dds_subset_grouped,contrast=c("group","totherKO","totherWT"),alpha=0.05))
res_global_tother<-as.data.frame(res_global_tother_raw)
res_global_tother$symbol<-mcols(dds_subset_grouped)$symbol
stopifnot(all.equal(res_global_tother$symbol,symbols[match(rownames(res_global_tother),symbols$gene_id),"gene_name"]))
#nrow(res_global_tother<-subset(res_global_tother,abs(log2FoldChange) > log2(1.5) & padj <0.05 & baseMean > 10)) # subset to significant
#res_global_tother<-res_global_tother[with(res_global_tother,order(padj)),] #sort based on padj
#head(as.data.frame(res_global_tother),10)
```

## Combine results from different cell types
```{r combine_cell_subtypes,eval=T}
#length(expressed_genes<-rownames(subset(results(dds_subset_grouped),baseMean > 10)))
#length(global_degs<-unique(c(rownames(res_global_th17), rownames(res_global_tfh), rownames(res_global_tfhth17), rownames(res_global_tother))))

res_global_th17_raw<-res_global_th17_raw[,c("baseMean","log2FoldChange","padj")]
colnames(res_global_th17_raw)<-c("baseMean","log2FoldChange.th17","padj.th17")

res_global_tfh_raw<-res_global_tfh_raw[,c("log2FoldChange","padj")]
colnames(res_global_tfh_raw)<-paste0(colnames(res_global_tfh_raw),".tfh")

res_global_tfhth17_raw<-res_global_tfhth17_raw[,c("log2FoldChange","padj")]
colnames(res_global_tfhth17_raw)<-paste0(colnames(res_global_tfhth17_raw),".tfhth17")

res_global_tother_raw<-res_global_tother_raw[,c("log2FoldChange","padj")]
colnames(res_global_tother_raw)<-paste0(colnames(res_global_tother_raw),".tother")

all.equal(rownames(res_global_th17_raw),rownames(res_global_tfh_raw))
all.equal(rownames(res_global_tfhth17_raw),rownames(res_global_tother_raw))
all.equal(rownames(res_global_tfhth17_raw),rownames(f_mean[rownames(res_global_tfhth17_raw),])) #add mean FPKM


global_results<-cbind(res_global_th17_raw,res_global_tfh_raw,res_global_tfhth17_raw,res_global_tother_raw,f_mean[rownames(res_global_tfhth17_raw),])
global_results$symbol<-mcols(dds_subset_grouped)$symbol
stopifnot(all.equal(global_results$symbol,symbols[match(rownames(global_results),symbols$gene_id),"gene_name"]))


dim(global_results<-global_results[global_results$baseMean > 0,])
length(expressed_genes<-rownames(global_results))

#global_results<-global_results[global_degs,]  #subset to deg in one or more group
```

## Four way Venn on global_results 
```{r four_way_venn,eval=T}
length(n1<-rownames(subset(global_results, padj.th17<0.05 & abs(log2FoldChange.th17) > 1)))
length(n2<-rownames(subset(global_results, padj.tfhth17<0.05 & abs(log2FoldChange.tfhth17) > 1)))
length(n3<-rownames(subset(global_results, padj.tfh<0.05 & abs(log2FoldChange.tfh) > 1)))
length(n4<-rownames(subset(global_results, padj.tother<0.05 & abs(log2FoldChange.tother) > 1)))


grid.newpage()
VennDiagram::draw.quad.venn(area1=length(n1),
                            area2=length(n2),
                            area3=length(n3),
                            area4=length(n4),
                            n12=length(intersect(n1,n2)),
                            n13=length(intersect(n1,n3)),
                            n14=length(intersect(n1,n4)),
                            n23=length(intersect(n2,n3)),
                            n24=length(intersect(n2,n4)),
                            n34=length(intersect(n3,n4)),
                            n123=length(intersect(intersect(n1,n2),n3)),
                            n124=length(intersect(intersect(n1,n2),n4)),
                            n134=length(intersect(intersect(n1,n3),n4)),
                            n234=length(intersect(intersect(n2,n3),n4)),
                            n1234=length(intersect(intersect(n1,n2),intersect(n3,n4))),
                            fill=c("#F8766D","#00BFC4","#7CAE00","#C77CFF"),
                            category=c("Th17","Th17-Tfh","Tfh","Other"))

global_results[intersect(n1,n2)[!intersect(n1,n2) %in% union(n3,n4)],]
```


```{r up_v_dn,eval=T}
#upregulated in 1 or more
length(u1<-rownames(subset(global_results, log2FoldChange.th17 > 1 & padj.th17<0.05 )))
length(u2<-rownames(subset(global_results, log2FoldChange.tfhth17 > 1 & padj.tfhth17<0.05 )))
length(u3<-rownames(subset(global_results, log2FoldChange.tfh > 1 & padj.tfh<0.05 )))
length(u4<-rownames(subset(global_results,log2FoldChange.tother > 1 & padj.tother<0.05 )))
nrow(global_results_up<-global_results[unique(c(u1,u2,u3,u4)),])

length(d1<-rownames(subset(global_results, log2FoldChange.th17 < -1 & padj.th17<0.05 )))
length(d2<-rownames(subset(global_results, log2FoldChange.tfhth17 < -1 & padj.tfhth17<0.05 )))
length(d3<-rownames(subset(global_results, log2FoldChange.tfh < -1 & padj.tfh<0.05 )))
length(d4<-rownames(subset(global_results,log2FoldChange.tother < -1 & padj.tother<0.05 )))
nrow(global_results_dn<-global_results[unique(c(d1,d2,d3,d4)),])


```

## Apply FPKM filter 
```{r fpkm_filter,eval=T}
# strong up based on FPKMs in KO > 2.5
length(su1<-rownames(subset(global_results, log2FoldChange.th17 > 1 & padj.th17<0.05 & fpkm_th17_KO > 2.5)))
length(su2<-rownames(subset(global_results, log2FoldChange.tfhth17 > 1 & padj.tfhth17<0.05 & fpkm_tfhth17_KO > 2.5)))
length(su3<-rownames(subset(global_results, log2FoldChange.tfh > 1 & padj.tfh<0.05 & fpkm_tfh_KO > 2.5)))
length(su4<-rownames(subset(global_results,log2FoldChange.tother > 1 & padj.tother<0.05 & fpkm_tother_KO > 2.5)))
nrow(global_results_strong_up<-global_results[unique(c(su1,su2,su3,su4)),])

#strong dn based on FPKMs in KO > 2.5
length(sd1<-rownames(subset(global_results, log2FoldChange.th17 < -1 & padj.th17<0.05 & fpkm_th17_WT > 2.5)))
length(sd2<-rownames(subset(global_results, log2FoldChange.tfhth17 < -1 & padj.tfhth17<0.05 & fpkm_tfhth17_WT > 2.5)))
length(sd3<-rownames(subset(global_results, log2FoldChange.tfh < -1 & padj.tfh<0.05 & fpkm_tfh_WT > 2.5)))
length(sd4<-rownames(subset(global_results,log2FoldChange.tother < -1 & padj.tother<0.05 & fpkm_tother_WT > 2.5)))
nrow(global_results_strong_dn<-global_results[unique(c(sd1,sd2,sd3,sd4)),])

#counts<-counts(dds_subset_grouped,normalized=TRUE)
#f<-fpkm(dds_subset_grouped,robust=T)
#nrow(subset(global_results_subset,baseMean > 100))
#global_results_subset[goiDF$gene_id,]
#counts["ENSMUSG00000031530",]
```


## Run GO analysis on UP and DN regulated Genes
```{r mgi_go,eval=T}
length(bd<-bd[names(bd) %in% expressed_genes])
head(bd)

#global_results_up
degs_global_results_up<-as.numeric(expressed_genes %in% rownames(global_results_up) )
names(degs_global_results_up)<-expressed_genes
table(degs_global_results_up)
go_wall_global_results_up<-goseq(nullp(degs_global_results_up,bias.data=bd),gene2cat=mgi_go.list[expressed_genes])
go_wall_global_results_up[1:20,c("category","term","over_represented_pvalue","numDEInCat","numInCat")]

#global_results_dn
degs_global_results_dn<-as.numeric(expressed_genes %in% rownames(global_results_dn) )
names(degs_global_results_dn)<-expressed_genes
table(degs_global_results_dn)
go_wall_global_results_dn<-goseq(nullp(degs_global_results_dn,bias.data=bd),gene2cat=mgi_go.list[expressed_genes])
go_wall_global_results_dn[1:20,c("category","term","over_represented_pvalue","numDEInCat","numInCat")]


#strong up
degs_global_strong_up<-as.numeric(expressed_genes %in% rownames(global_results_strong_up)) 
names(degs_global_strong_up)<-expressed_genes
table(degs_global_strong_up)
go_wall_global_strong_up <-goseq(nullp(degs_global_strong_up,bias.data=bd),gene2cat=mgi_go.list[expressed_genes])
go_wall_global_strong_up[1:20,c("category","term","over_represented_pvalue","numDEInCat","numInCat")]

#strong dn
degs_global_strong_dn<-as.numeric(expressed_genes %in% rownames(global_results_strong_dn)) 
names(degs_global_strong_dn)<-expressed_genes
table(degs_global_strong_dn)
go_wall_global_strong_dn <-goseq(nullp(degs_global_strong_dn,bias.data=bd),gene2cat=mgi_go.list[expressed_genes])
go_wall_global_strong_dn[1:20,c("category","term","over_represented_pvalue","numDEInCat","numInCat")]

#figure
GO_UP<-head(go_wall_global_strong_up,20) %>%
  dplyr::mutate(term=factor(term,levels=rev(term))) %>%
ggplot(aes(x=term,y=-log10(over_represented_pvalue),size=numDEInCat)) +
  geom_point() + ylim(c(0,14))+
 # geom_bar(stat="identity",fill="red") +
  coord_flip() + xlab("") +
  theme_bw() 

GO_DN<-head(go_wall_global_strong_dn,20) %>%
  dplyr::mutate(term=factor(term,levels=rev(term))) %>%
ggplot(aes(x=term,y=-log10(over_represented_pvalue),size=numDEInCat)) +
  geom_point() + ylim(c(0,14))+
 # geom_bar(stat="identity",fill="red") +
  coord_flip() + xlab("") +
  theme_bw() 

grid.arrange(GO_UP,GO_DN,ncol=1)


#interesting gategories
listGO("GO:0007186",dg=degs_global_strong_up) #Ramp1, Ccr, Cxcr
listGO("GO:2000320",dg=degs_global_strong_up) # Foxp3
listGO("GO:0043433",dg=degs_global_strong_up) # Foxp3, Id2, Smad7, Sirt1
listGO("GO:0007165",dg=degs_global_strong_up) #Ptch1
listGO("GO:2000320",degs_global_strong_up) #Smad7, Tbx21

listGO("GO:0006260",dg=degs_global_strong_dn) # Blm, Pola/d/q
listGO("GO:0006270",dg=degs_global_strong_dn) # Mcm/Orc

```


# ChIP Analysis

## Download Peaklist
```{r download_peaklist,eval=F}
#ChIP Tracks
#https://s3.msi.umn.edu/jenkinsm/th17_input_P038_Index7_S18.fastq.bigWig
#https://s3.msi.umn.edu/jenkinsm/th17_BCOR_P33P36_combined.fastq.bigWig

length(th17n_Bcor_peaks<-narrowPeakToGRanges(url("https://s3.msi.umn.edu/bardwell/th17_Bcor_WT_113017_peaks.narrowPeak"),name="th17")) 
length(th17b_Bcor_peaks<-broadPeakToGRanges(url("https://s3.msi.umn.edu/jenkinsm/th17_bcor_083117_peaks.broadPeak"),name="th17")) 


length(rorgt_peaks_rep1<-narrowPeakToGRanges(url("https://s3.msi.umn.edu/jenkinsm/rorg_th17_rep1_100418_peaks.narrowPeak"),name="rorgt")) 
length(rorgt_peaks_rep2<-narrowPeakToGRanges(url("https://s3.msi.umn.edu/jenkinsm/rorg_th17_rep2_100418_peaks.narrowPeak"),name="rorgt")) 

length(cgi<-narrowPeakToGRanges(url("https://s3.msi.umn.edu/kybam/cpg_mm10.bed"),name="cgi",skip=1)) 
seqlevelsStyle(cgi)<-"NCBI"
cgi<-keepStandardChromosomes(cgi,pruning.mode="coarse")

#save Peaklists for Offline Analysis  
save(th17n_Bcor_peaks,th17b_Bcor_peaks,cgi,file="th17_peaklists.rdata")
```


## Annotate Peaks & Genes

```{r annotate,eval=T}
load("th17_peaklists.rdata")

#How many Overlap CpG Islands?

length(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103])  # q< 0.05
sum(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103] %over% cgi)
sum(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103] %over% cgi)/length(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103])
sum(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103] %over% prom)
sum(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103] %over% prom)/length(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103])

prom_cgi<-prom[prom %over% cgi]
sum(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103] %over% prom_cgi)/length(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103])

  # Annotate Peaks
head(th17b_Bcor_peaks_anno<-annotate_peaks(th17b_Bcor_peaks,tx=transcripts[transcripts$gene_id %in% expressed_genes],cutoff=5000))

#head(global_results<-annotate_genes(df=global_results,gr=th17b_Bcor_peaks,tx=transcripts[transcripts$gene_id %in% expressed_genes],cutoff=10000))
#View(global_results)

#Annotate 157 Upregulated Genes
head(global_results_up_chip<-annotate_genes(df=global_results_up,gr=th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103],tx=transcripts[transcripts$gene_id %in% expressed_genes],cutoff=5000))
#write.csv(tibble::rownames_to_column(as.data.frame(global_results),var="Ensembl_Gene_ID"),file=paste0("global_results_with_ChIP_annotation_",ts,".csv"),quote=F,row.names = F)

# How many bound by BCOR
sum(!is.na(global_results_up_chip$maxscore))

# strong chip up
length(scu1<-rownames(subset(global_results_up_chip, log2FoldChange.th17 > 1 & padj.th17<0.05  & !is.na(maxscore))))
length(scu2<-rownames(subset(global_results_up_chip, log2FoldChange.tfhth17 > 1 & padj.tfhth17<0.05 &  !is.na(maxscore))))
length(scu3<-rownames(subset(global_results_up_chip, log2FoldChange.tfh > 1 & padj.tfh<0.05  & !is.na(maxscore))))
length(scu4<-rownames(subset(global_results_up_chip,log2FoldChange.tother > 1 & padj.tother<0.05  & !is.na(maxscore))))
nrow(global_results_strong_chip_up<-global_results[unique(c(scu1,scu2,scu3,scu4)),])

#goiDF$gene_id %in% rownames(global_results_strong_chip_up)

```





## Export Direct Targets to CSV and PDF
```{r Supplemental_Table_1,eval=T}


temp<-tibble::rownames_to_column(as.data.frame(global_results_strong_chip_up),var="Ensembl Gene Id")

temp<-temp[,c("Ensembl Gene Id","symbol" ,"fpkm_th17_WT","fpkm_th17_KO","log2FoldChange.th17","padj.th17",
                                          "fpkm_tfhth17_WT","fpkm_tfhth17_KO","log2FoldChange.tfhth17","padj.tfhth17" ,
                                          "fpkm_tfh_WT", "fpkm_tfh_KO","log2FoldChange.tfh","padj.tfh",
                                          "fpkm_tother_WT" ,  "fpkm_tother_KO","log2FoldChange.tother", "padj.tother")] 

temp$log2FoldChange.th17<-round(temp$log2FoldChange.th17,3)
temp$padj.th17<-round(temp$padj.th17,3)

temp$log2FoldChange.tfhth17<-round(temp$log2FoldChange.tfhth17,3)
temp$padj.tfhth17<-round(temp$padj.tfhth17,3)

temp$log2FoldChange.tfh<-round(temp$log2FoldChange.tfh,3)
temp$padj.tfh<-round(temp$padj.tfh,3)

temp$log2FoldChange.tother<-round(temp$log2FoldChange.tother,3)
temp$padj.tother<-round(temp$padj.tother,3)

colnames(temp)<-c("Ensembl Gene Id","Gene Symbol" ,"FPKM Th17 WT","FPKM Th17 KO","Log2FoldChange Th17","padj Th17",
                                          "FPKM Th17-Tfh_WT","FPKM Th17-Tfh KO","Log2FoldChange Th17-Tfh","padj Th17-Tfh" ,
                                          "FPKM Tfh WT", "FPKM Tfh KO","Log2FoldChange Tfh","Padj Tfh",
                                          "FPKM RORgt-CXCR5- WT",  "FPKM RORgt-CXCR5- KO", "log2FoldChange RORgt-CXCR5-", "padj RORgt-CXCR5-")


write.csv(temp,file=paste0("Supplemental_Table_1_",ts,".csv"),quote=F,row.names = F)
```


## Make a Latex Version of Supplmental Table 1
```{r tex,eval=F}
temp1<-temp[,c("Ensembl Gene Id","symbol","fpkm_th17_WT","fpkm_th17_KO", "fpkm_tfhth17_WT","fpkm_tfhth17_KO",
                                          "fpkm_tfh_WT", "fpkm_tfh_KO", "fpkm_tother_WT" ,  "fpkm_tother_KO")] 
colnames(temp1)<-c("Ensembl Gene Id","Gene Symbol" ,"Th17 WT","Th17 KO", "Th17-Tfh_WT","Th17-Tfh KO",
                                          "Tfh WT", "Tfh KO", "RORgt-CXCR5- WT", "RORgt-CXCR5- KO")

temp2<-temp[,c("Ensembl Gene Id","symbol","log2FoldChange.th17","padj.th17", "log2FoldChange.tfhth17","padj.tfhth17" ,
               "log2FoldChange.tfh","padj.tfh", "log2FoldChange.tother", "padj.tother")] 

colnames(temp2)<-c("Ensembl Gene Id","Gene Symbol" ,"FPKM Th17 WT","FPKM Th17 KO","Log2FoldChange Th17","padj Th17",
                                          "FPKM Th17-Tfh_WT","FPKM Th17-Tfh KO","Log2FoldChange Th17-Tfh","padj Th17-Tfh" ,
                                          "FPKM Tfh WT", "FPKM Tfh KO","Log2FoldChange Tfh","Padj Tfh",
                                          "FPKM RORgt-CXCR5- WT",  "FPKM RORgt-CXCR5- KO", "log2FoldChange RORgt-CXCR5-", "padj RORgt-CXCR5-")


xtmp1 <- xtable(temp1, caption = "A \\code{longtable} spanning several pages")
print(xtmp1, file="temp.tex", hline.after=c(-1, 0), tabular.environment = "longtable",floating=FALSE,include.rownames=FALSE)

xtmp1<-xtable(temp1)
#digits(xtmp1)<-c(0,0,2,-2)
print(xtmp1, file="temp.tex", floating=FALSE,include.rownames=FALSE)
system("pdflatex test6.tex")
system(paste0("mv test6.pdf ","Supplmentary_Table_1_",ts,".pdf"))
```

## Identify candidate Th17 drivers based on ontology and KO counts
```{r find_candidates,eval=T}


#Add Count Data to global_results_up_chip
global_results_strong_chip_up_counts<-cbind(global_results_strong_chip_up,count_mean[rownames(global_results_strong_chip_up),])
#head(global_results_strong_chip_up_counts)

#subtype_specific_results<-subtype_specific_results[rownames(subtype_specific_results) %in% rownames(global_results_strong_chip_up),]
#View(as.data.frame(global_results_strong_chip_up_counts))

length(ssu1<-rownames(subset(global_results_strong_chip_up_counts, log2FoldChange.th17 > 1 & padj.th17<0.05 & count_th17_KO > 100)))
length(ssu2<-rownames(subset(global_results_strong_chip_up_counts, log2FoldChange.tfhth17 > 1 & padj.tfhth17<0.05  & count_tfhth17_KO > 100)))
length(ssu3<-rownames(subset(global_results_strong_chip_up_counts, log2FoldChange.tfh > 1 & padj.tfh<0.05 & count_tfh_KO > 100)))
length(ssu4<-rownames(subset(global_results_strong_chip_up_counts, log2FoldChange.tother > 1 & padj.tother < 0.05 & count_tother_KO > 100 )))
nrow(candidate_genes<-global_results_strong_chip_up_counts[unique(c(ssu1,ssu2,ssu3,ssu4)),])
#nrow(strong_up<-subset(strong_up,baseMean.th17 > 100 | baseMean.tfhth17 > 100 | baseMean.tfh > 100 | baseMean.tother > 100))

View(as.data.frame(candidate_genes))
goiDF$gene_id %in% rownames(candidate_genes)
#goiDF[!goiDF$gene_id %in% rownames(candidate_genes),]

candidate_genes$ontology<-sapply(rownames(candidate_genes),geneToGOIDs)

nrow(candidate_genes_onto<-candidate_genes[str_detect(tolower(candidate_genes$ontology),"transcription|phosphorylation"),])
goiDF$gene_id %in% rownames(candidate_genes_onto)

#View(as.data.frame(candidate_genes_onto[!rownames(candidate_genes_onto) %in% goiDF$gene_id,]))
```

## Heatmap for Candidate Drivers
```{r goi_heatmap,eval=T}

goi<-c("Lef1","Runx2","Id3","Foxo3","Bach2","Dusp4","Dusp6","Hipk2","Tsc22d3","Scml4","Ptch1")  #Add "Scml4" and remove c("Igf1r","Smad7","Ccnd2","Pip4k2a")
goiDF<-symbols[symbols$gene_name %in% goi,]
rownames(goiDF)<-goiDF$gene_name
goiDF<-goiDF[goi,] #put in the original order
goiDF$chip_score<-c(75,63,42,73,35,43,72,80,32,13,67)
goiDF$ontology<-c(rep("Transcription Factor",5),rep("Phosphatase",2),"Serine/Threonine Kinase","Leucine Zipper Protein","Polycomb Group Protein","Sonic Hedgehog Receptor")


dim(f_mean_log2<-log2(f_mean[goiDF$gene_id,]))
rownames(f_mean_log2)<-symbols[match(rownames(f_mean_log2),symbols$gene_id),"gene_name"]

temp2<-as.matrix(t(scale(t(f_mean_log2),center=T,scale=T)))

# Make Heatmap
# colors http://colorbrewer2.org/#type=qualitative&scheme=Dark2&n=6

col_fun = circlize::colorRamp2(c(min(temp2[,1:8]), (min(temp2[,1:8])+max(temp2[,1:8]))/2, max(temp2[,1:8])), c('#ffeda0','#feb24c','#f03b20'))
col_fun2 = circlize::colorRamp2(c(0,50,100), c('#f7fcb9','#addd8e','#31a354'))

all.equal(rownames(temp2),rownames(goiDF)) #check to make sure temp2 is in the right order
temp2<-cbind(temp2,goiDF[,c("chip_score","ontology")])

a<-Heatmap(temp2[,c(which(str_detect(colnames(temp2),"^th17_WT")),which(str_detect(colnames(temp2),"^th17_KO")))], 
                  column_title = "th17", rect_gp = gpar(col = "black", lwd = 0.5),
                  cluster_rows = T, cluster_columns=FALSE,
                  row_names_side = "left",col = col_fun, show_heatmap_legend = FALSE)
b<-Heatmap(temp2[,c(which(str_detect(colnames(temp2),"^tfhth17_WT")),which(str_detect(colnames(temp2),"^tfhth17_KO")))], 
                  column_title = "tfhth17", rect_gp = gpar(col = "black", lwd = 0.5),show_row_names=F,
                  cluster_rows = T, cluster_columns=FALSE,
                  row_names_side = "left",col = col_fun, show_heatmap_legend = FALSE)
c<-Heatmap(temp2[,c(which(str_detect(colnames(temp2),"^tfh_WT")),which(str_detect(colnames(temp2),"^tfh_KO")))], 
                  column_title = "tfh", rect_gp = gpar(col = "black", lwd = 0.5),show_row_names=F,
                  cluster_rows = T, cluster_columns=FALSE,
                  row_names_side = "left",col = col_fun, show_heatmap_legend = FALSE)
d<-Heatmap(temp2[,c(which(str_detect(colnames(temp2),"^tother_WT")),which(str_detect(colnames(temp2),"^tother_KO")))], 
                  column_title = "tother", rect_gp = gpar(col = "black", lwd = 0.5),show_row_names=F,
                  cluster_rows = T, cluster_columns=FALSE, name = "Centered,\nScaled\nLog2 \nFPKM",
                  row_names_side = "left",col = col_fun, show_heatmap_legend = TRUE)

e<- Heatmap(temp2[,which(str_detect(colnames(temp2),"chip")),drop=F], name = "ChIP\nPeak\nScore",width = unit(0.5, "cm"),
            rect_gp = gpar(col = "black", lwd = 0.5),col = col_fun2,show_row_names=F)
f<-rowAnnotation(df = temp2[,which(str_detect(colnames(temp2),"ontology")),drop=F], 
                 col = list(ontology = c("Transcription Factor" = '#1b9e77',
                                     "Phosphatase" = '#d95f02',
                                     "Serine/Threonine Kinase"='#7570b3',
                                     "Leucine Zipper Protein"='#e7298a',
                                     "Polycomb Group Protein"='#e41a1c',
                                     "Sonic Hedgehog Receptor"='#e6ab02'))) 
draw(a+b+c+d+e+f)
```

## Exon 9 & 10 Deletion Efficiency

```{r exon910,eval=T}

symbols[grep("Bcor",symbols$gene_name),]
as.data.frame(transcripts[grep("Bcor",transcripts$symbol),])
gr<-exons[["ENSMUST00000115513.8"]][7:10]
strand(gr)<-"*"
export(gr,"exons78910.bed")

rtracklayer::import("../data/tfh_KO_rep1_P041_Pool1_Index8_q0_scaled_fwd.bigWig",selection=gr)
files_known<-list.files("../data","*.bigWig")

files<-c(paste0(colData(dds_subset)$sample_name[1:22],"_q0_scaled_fwd.bigWig"),paste0(colData(dds_subset)$sample_name[23:36],".bigWig"))
bw<-BigWigFileList(paste0("../data/",files))
names(bw)<-colData(dds_subset)$sample_name
str(bw)
#bw<-BigWigFileList("../data/tfh_KO_rep1_P041_Pool1_Index8_q0_scaled_fwd.bigWig")

#http://lcolladotor.github.io/protocols/bigwig_DEanalysis/ (thanks to L. Collado-Torres) 
counts <- matrix(NA, nrow = length(gr), ncol = length(bw))
colnames(counts) <- names(bw)
for(i in seq_len(length(bw))) {
    coverage <- import(bw[[i]], as = 'RleList')$X
    counts[, i] <- sum(Views(coverage, ranges(gr)))
}
## Divide by read length and round to integer numbers
counts <- round(counts / 50, 0)

## Explore a little portion of the count matrix 
dim(counts)
spl<-as.data.frame(t(counts))
spl$ratio<-(spl[,3]+spl[,4])/(spl[,1]+spl[,2])
spl$celltype<-colData(dds_subset)$celltype
spl$genotype<-colData(dds_subset)$genotype
spl<-spl[,c("ratio","celltype","genotype")]

spl %>% 
  mutate(celltype=factor(celltype,levels=c("th17","tfhth17","tfh","tother"))) %>% 
  ggplot(aes(x=genotype,y=ratio))+
    geom_point(position=position_jitter(w=0.1,h=0)) + ggtitle("Count Ratios for Exons 9 & 10 / Exons 7 & 8 ") +
    stat_summary(fun.y = mean, fun.ymin = mean, fun.ymax = mean,
                 geom = "crossbar", width = 0.35,color="blue",size=0.2) +
    expand_limits(x=0, y = 0) + xlab("") +
    facet_grid(~celltype) +
    theme(legend.position="none") +
    theme_bw()
```



ChIP Enrichment
```{r blah4,eval=F}
length(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103])
sum(th17b_Bcor_peaks[th17b_Bcor_peaks$q > 1.30103] %over% cgi)
length(cgi)-sum(th17b_Bcor_peaks %over% cgi)
length(th17b_Bcor_peaks)-sum(th17b_Bcor_peaks %over% cgi)
library("BSgenome.Mmusculus.UCSC.mm10.masked")
mm10<-BSgenome.Mmusculus.UCSC.mm10.masked
seqlevelsStyle(mm10)<-"NCBI"

length(cgi<-keepStandardChromosomes(cgi,pruning.mode = "coarse"))
length(th17b_Bcor_peaks<-keepStandardChromosomes(th17b_Bcor_peaks,pruning.mode = "coarse"))
length(rorgt_peaks_rep1<-keepStandardChromosomes(rorgt_peaks_rep1,pruning.mode = "coarse"))

numOverlaps(th17b_Bcor_peaks, cgi, count.once=TRUE)
numOverlaps(randomizeRegions(th17b_Bcor_peaks,genome=mm10), cgi, count.once=TRUE)

(pt <- overlapPermTest(A=th17b_Bcor_peaks, B=cgi, ntimes=20,genome=mm10,non.overlapping=TRUE))
pt2 <- overlapPermTest(A=th17b_Bcor_peaks, B=rorgt_peaks_rep1, ntimes=50,genome=mm10)
pt3 <- overlapPermTest(A=th17b_Bcor_peaks, B=rorgt_peaks_rep2, ntimes=20,genome=mm10)
                      
pt4 <- overlapPermTest(A=rorgt_peaks_rep1, B=cgi, ntimes=20,genome=mm10)
```


```{r blah5,eval=F}
x<-apply(global_results[,10:17],1,sum)
length(expressed_genes<-names(x[x>0]))
length(expressed_genes<-rownames(dds))

length(bcor_targets<-peaks_to_genes(th17b_Bcor_peaks,tx=transcripts[transcripts$gene_id %in% expressed_genes],cutoff=1000))
length(cpg_targets<-peaks_to_genes(cgi,tx=transcripts[transcripts$gene_id %in% expressed_genes],cutoff=10000))
length(rorgt_targets<-peaks_to_genes(rorgt_peaks_rep1,tx=transcripts[transcripts$gene_id %in% expressed_genes],cutoff=10000))
length(expressed_genes)

grid.newpage()
draw.pairwise.venn(length(bcor_targets),length(cpg_targets),sum(bcor_targets %in% cpg_targets),
                   fill=c("green","cyan"),
                   category=c("BCOR targets","Cpg Islands"),
                   inverted=F)

(m<-matrix(c(sum(bcor_targets %in% cpg_targets), 
         length(bcor_targets)-sum(bcor_targets %in% cpg_targets),
         length(cpg_targets)-sum(bcor_targets %in% cpg_targets), 
         length(expressed_genes)-length(bcor_targets)-length(cpg_targets)+sum(bcor_targets %in% cpg_targets)),
         2, 2, dimnames = list(cpg_targets = c("true", "false"), bcor_targets = c("true", "false"))))
fisher.test(m)$p.value

grid.newpage()
draw.pairwise.venn(length(bcor_targets),length(rorgt_targets),sum(bcor_targets %in% rorgt_targets),
                   fill=c("green","cyan"),
                   category=c("BCOR targets","RORgt Islands"),
                   inverted=F)

(m<-matrix(c(sum(bcor_targets %in% rorgt_targets), 
         length(bcor_targets)-sum(bcor_targets %in% rorgt_targets),
         length(rorgt_targets)-sum(bcor_targets %in% rorgt_targets), 
         length(expressed_genes)-length(bcor_targets)-length(rorgt_targets)+sum(bcor_targets %in% rorgt_targets)),
         2, 2, dimnames = list(rorgt_targets = c("true", "false"), bcor_targets = c("true", "false"))))
fisher.test(m,workspace=30000000)$p.value

```

#promoters
```{r blah6,eval=F}
prom_bcor<-prom[prom %over% th17b_Bcor_peaks]
prom_cgi<-prom[prom %over% cgi]
prom_157<-prom[prom$gene_id %in% rownames(global_results_up)]
length(unique(prom_157$gene_id))
length(unique(prom_157[ prom_157 %over% th17b_Bcor_peaks]$gene_id))
length(unique(prom_157[prom_157 %over% cgi & prom %over% th17b_Bcor_peaks]$gene_id))


length(unique(prom_bcor$symbol))
goiDF$bound<-goiDF$gene_id %in% prom_bcor$gene_id
goiDF$cgi<-goiDF$gene_id %in% prom_cgi$gene_id
goiDF

```

# SessionInfo
```{r sessioninfo,eval=T}
sessionInfo()
```

